src/features/klant/service.ts

```diff
@@ -52,13 +52,31 @@ const identificatorTypes = {
   // },
 } as const satisfies Record<string, IdentificatorType>;
 
-const partijTypes = {
+type Contactnaam = {
+  voornaam: string;
+  voorvoegselAchternaam?: string;
+  achternaam: string;
+};
+
+type PartijTypes = {
+  persoon: {
+    contactnaam: Contactnaam;
+  };
+  organisatie: {
+    naam: string;
+  };
+  contactpersoon: {
+    contactnaam: Contactnaam;
+  };
+};
+
+type PartijType = keyof PartijTypes;
+
+const partijTypes: { [k in PartijType]: k } = {
```

Comment:
behoorlijk hardcore met dit soort constructies `{ [k in PartijType]: k }` en `const createKlant = <K extends PartijType>(soortPartij: K, partijIdentificatie: PartijTypes[K]`. 

Wordt er niet per se leesbaarder op voor een niet heel senior ts developer. waarom niet zoiets?
 
```
type PersoonIdentificator = {
  contactnaam: Contactnaam;
};

type OrganisatieIdentificator = {
  naam: string;
};

type ContactpersoonIdentificator = {
  contactnaam: Contactnaam;
};

type PartijIdentificator =
  | PersoonIdentificator
  | OrganisatieIdentificator
  | ContactpersoonIdentificator;

enum PartijTypes {
  persoon = "persoon",
  organisatie = "organisatie",
  contactpersoon = "contactpersoon",
}

... en dan 

const createKlant = (
  soortPartij: PartijTypes,
  partijIdentificatie: PartijIdentificator,
  identificatorType: IdentificatorType,
  objectId: string,
) => ...

.. en bij ensureklantForBsn weet je al dat het om een persoon gaat, dus ipv

ensureKlantForBsn({
  bsn,
  partijIdentificatie,
}: {
  bsn: string;
  partijIdentificatie: PartijTypes["persoon"];
}) { ...

volstaat dan dit. (oogt naar mijn idee wat eenvoudiger en toegankelijker) 

ensureKlantForBsn(
  bsn: string,
  partijIdentificatie: PersoonIdentificator,
) { ...

```

nb. ik heb dit niet helemaal uitgedacht. ik zie misschien iets over het hoofd. het gaat meer om het principe. Kijk maar of je er wat mee kan

Reply:
Ik heb de strekking van je voorbeeld overgenomen: enum  en sum types in plaats van zo'n generic constructie.